### PR TITLE
[fixes egg-200]: Display different components based on sanity property on Home page

### PR DESCRIPTION
**Pull request by zacjones93**
The homepage is a dynamically rendered list of sections based on what the page receives from Sanity. Up to this point we only displayed one type of component, a section that displayed a header, some text and a dynamic grid of resource cards. 

This PR updates the home page behavior to allow for us to control what component is used for a section on the page. 

We want to show an "egghead for teams CTA" so this PR adds the first custom display component `EggheadForTeamsCTA`. The default will be the existing section behavior that we already see on the home page.

## DynamicGrid Component:
Dynamic grid component

## EggheadForTeamsCTA

EggheadForTeamsCTA dark desktop
EggheadForTeamsCTA light desktop
EggheadForTeamsCTA dark mobile
EggheadForTeamsCTA light mobile

### CODE REVIEW

File: src/components/pages/home/egghead-for-teams-cta.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                <div className="absolute top-0 left-[14px] h-2 w-2 bg-amber-500 after:-z-10"></div>
+                <div className="absolute right-0 bottom-[14px] h-2 w-2 bg-amber-500"></div>
+                <a
+                  className="absolute bottom-16 -right-12 rotate-45  w-[141.42%] rounded-sm px-5 py-2 
+              text-sm font-semibold uppercase tracking-wide text-center shadow-lg text-amber-800
+              bg-gradient-to-br from-amber-300 via-amber-100 to-yellow-300 animate-gradient-xy hover:from-amber-400 hover:via-amber-200 hover:to-yellow-400"
+                >
+                  For Teams
+                </a>
+              </div>
```
Comment (zacjones93) [inline src/components/pages/home/egghead-for-teams-cta.tsx:23]: Banner on the top right

File: src/components/pages/home/egghead-for-teams-cta.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import * as React from 'react'
+import {FunctionComponent} from 'react'
+import Image from 'next/image'
+import Link from 'next/link'
+
+const EggheadForTeamsCta: FunctionComponent = () => {
+  return (
+    <section className=" mb-16">
+      <Link href="/egghead-for-teams">
```
Comment (zacjones93) [inline src/components/pages/home/egghead-for-teams-cta.tsx:9]: The whole card is a link

File: src/components/pages/home/index.tsx [hunk 3]
```diff
@@ -47,143 +46,11 @@ const Home: React.FC<any> = ({data, jumbotron, location}) => {
...
-                    </div>
-                  )}
-                </section>
-              )
-            })}
+          <SanitySections
+            sections={data.sections}
+            location={location}
+            completedCoursesIds={completedCoursesIds}
+          />
```
Comment (zacjones93) [inline src/components/pages/home/index.tsx:53]: most of the home page is just the `SanitySections` which is extracted to it's own component

File: src/components/pages/home/sanity-sections.tsx [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import ReactMarkdown from 'react-markdown'
+import {CardResource} from 'types'
+import Grid from 'components/grid'
+import Image from 'next/image'
+import Link from 'next/link'
+import EggheadForTeamsCta from 'components/pages/home/egghead-for-teams-cta'
+import analytics from 'utils/analytics'
+import {SanitySectionType} from 'pages/learn'
+
+const DynamicGridComponent = ({
```
Comment (zacjones93) [inline src/components/pages/home/sanity-sections.tsx:14]: `DyncamicGridComponent` extracted into it's own component.

File: src/components/pages/home/sanity-sections.tsx [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+          return (
+            <DynamicGridComponent
+              section={section}
+              key={i}
+              location={location}
+              completedCoursesIds={completedCoursesIds}
+            />
+          )
+        })}
```
Comment (zacjones93) [inline src/components/pages/home/sanity-sections.tsx:183]: switch on topics (that was behavior we already had in the home page so left it), and `section.displayComponent`. 

We could add more `displayComponent` types in sanity if we wanted to customize sections.

Default is the `DynamicGridComponent` we are used to.

File: src/pages/learn/index.tsx [hunk 6]
```diff
@@ -79,8 +85,65 @@ const homepageQuery = groq`*[_type == 'resource' && slug.current == "curated-hom
     }
   }`
 
+const Resource = z.object({
```
Comment (zacjones93) [inline src/pages/learn/index.tsx:88]: typed the resource coming in from sanity

File: src/pages/learn/index.tsx [hunk 7]
```diff
@@ -79,8 +85,65 @@ const homepageQuery = groq`*[_type == 'resource' && slug.current == "curated-hom
...
+  title: z.string(),
+  sections: z.array(SanitySection).nonempty(),
+})
+export type CuratedHomePageDataType = z.infer<typeof CuratedHomePageData>
+
 export async function getStaticProps() {
-  const data = await sanityClient.fetch(homepageQuery)
+  const data = await sanityClient
+    .fetch(homepageQuery)
+    .then((result) => CuratedHomePageData.parse(result))
```
Comment (zacjones93) [inline src/pages/learn/index.tsx]: parsing the data coming in through zod

File: src/types.ts [hunk 8]
```diff
@@ -1,21 +1,21 @@
...
-  title: string
-  slug: string
-  description: string
+  title?: string
+  slug?: string
+  description?: string
   icon_url?: string
   completed?: boolean
-  path: string
+  path?: string
```
Comment (zacjones93) [inline src/types.ts:7]: adjusted types to match sanity data coming in

File: studio/schemas/documents/resource.js [hunk 9]
```diff
@@ -125,6 +125,27 @@ export default {
         ],
       },
     },
+    {
+      name: 'displayComponent',
+      type: 'string',
+      title: 'Display Component',
+      description:
+        'What component should be used in egghead-next to display this feature?',
+      // hidden: ({document}) => document.type !== 'feature',
```
Comment (zacjones93) [inline studio/schemas/documents/resource.js:134]: @Creeland hiding this field based on document type wasn't working in nested resources... not sure if this was a bug or something that I was doing wrong.

File: studio/schemas/documents/resource.js [hunk 10]
```diff
@@ -125,6 +125,27 @@ export default {
...
+      options: {
+        list: [
+          {
+            title: 'Dynamic Grid',
+            value: 'DynamicGrid',
+            default: true,
```
Comment (zacjones93) [inline studio/schemas/documents/resource.js:140]: `default: true` also didn't do anything though on the front end this is already a default so we could probably just remove this option? If we don't set any custom component the dynamic grid shows

File: tailwind.config.js [hunk 11]
```diff
@@ -124,6 +124,43 @@ module.exports = {
...
+          '0%, 100%': {
+            'background-size': '400% 400%',
+            'background-position': 'left center',
+          },
+          '50%': {
+            'background-size': '200% 200%',
+            'background-position': 'right center',
+          },
+        },
+      },
```
Comment (zacjones93) [inline tailwind.config.js:163]: added some animation for the eggheadForTeams CTA

**Review by joelhooks — APPROVED** [review]

**Pull request merged**